test/Microsoft.AspNet.Mvc.Core.Test/Routing/UrlHelperTest.cs

```diff
@@ -851,6 +857,210 @@ public void Action_RouteValueInvalidation_DoesNotAffectActionAndController_Actio
+        {
+            // Arrange
+            var services = GetServices();
+            var router = BuildRouterWithAttributeRoute(services);
+            var urlHelper = CreateUrlHelper("/", router, services);
+
+            // Act
+            var url = urlHelper.Action(
+                action: "Get",
+                controller: "MatchVariable");
+
+            // Assert
+            Assert.Equal("/api/MatchVariable/help", url);
+        }
+
+        private static IRouter BuildRouterWithAttributeRoute(IServiceProvider services)
+        {
+            var routeBuilder = new RouteBuilder
+            {
+                DefaultHandler = new MvcRouteHandler(),
+                ServiceProvider = services
+            };
+
+            routeBuilder.MapRoute("Default", "{controller}/{action}");
+
+            //Adding the attribute route as done in "UseMvc"
+            routeBuilder.Routes.Insert(0, AttributeRouting.CreateAttributeMegaRoute(
+                routeBuilder.DefaultHandler,
+                services));
+
+            return routeBuilder.Build();
+        }
+
+        private static IServiceProvider GetServices()
+        {
+            var invoker = new Mock<IActionInvoker>();
+            invoker
+                .Setup(i => i.InvokeAsync())
+                .Returns(Task.FromResult(true));
+
+            var invokerFactory = new Mock<IActionInvokerFactory>();
+            invokerFactory
+                .Setup(f => f.CreateInvoker(It.IsAny<ActionContext>()))
+                .Returns<ActionContext>((c) =>
+                {
+                    return invoker.Object;
+                });
+
+            var optionsAccessor = new Mock<IOptions<RouteOptions>>();
+            optionsAccessor
+                .SetupGet(o => o.Value)
+                .Returns(new RouteOptions());
+            var loggerFactory = NullLoggerFactory.Instance;
+
+            var serviceMock = new Mock<IServiceProvider>();
+            serviceMock
+                .Setup(s => s.GetService(typeof(IOptions<RouteOptions>)))
+                .Returns(optionsAccessor.Object);
+            serviceMock
+                .Setup(s => s.GetService(typeof(ILoggerFactory)))
+                .Returns(loggerFactory);
+
+            var actionContextAccessor = new ActionContextAccessor()
+            {
+                ActionContext = new ActionContext()
+                {
+                    HttpContext = new DefaultHttpContext()
+                    {
+                        ApplicationServices = serviceMock.Object,
+                        RequestServices = serviceMock.Object,
+                    },
+                    RouteData = new RouteData(),
+                },
+            };
+
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddMvc();
+
+            serviceCollection.AddInstance<ILibraryManager>(CreateLibraryManager());
+            serviceCollection.AddInstance<IActionInvokerFactory>(invokerFactory.Object);
```

Comment:
You've added a lot of services to this method aren't used in existing; also, there's services here that aren't needed for your tests. Makes me worry about conflating the test expectations.